File: ios/ShadowCell/Models/ThreatEvent.swift
import Foundation

enum ThreatEventType: String, Codable {
    case networkDowngrade       = "NETWORK_DOWNGRADE"
    case networkUpgradeFast     = "NETWORK_UPGRADE_FAST"
    case silentSmsCandidate     = "SILENT_SMS_CANDIDATE"
    case cellTowerAnomaly       = "CELL_TOWER_ANOMALY"
    case locationUpdateBurst    = "LOCATION_UPDATE_BURST"
    case signalDropHandoff      = "SIGNAL_DROP_HANDOFF"
    case pagingAnomaly          = "PAGING_ANOMALY"
}

enum RiskLevel: Int, Comparable, Codable {
    case safe     = 0
    case medium   = 1
    case high     = 2
    case critical = 3

    static func < (lhs: RiskLevel, rhs: RiskLevel) -> Bool { lhs.rawValue < rhs.rawValue }

    var label: String {
        switch self {
        case .safe:     return "GÜVENLI"
        case .medium:   return "ORTA"
        case .high:     return "YÜKSEK"
        case .critical: return "KRİTİK"
        }
    }

    var color: String {  // SwiftUI'da Color olarak kullanılacak
        switch self {
        case .safe:     return "green"
        case .medium:   return "yellow"
        case .high:     return "orange"
        case .critical: return "red"
        }
    }
}

struct ThreatEvent: Identifiable, Codable {
    let id: UUID
    let timestamp: Date
    let type: ThreatEventType
    let rawValue: String
    let score: Int
    let context: String

    init(
        id: UUID = UUID(),
        timestamp: Date = Date(),
        type: ThreatEventType,
        rawValue: String,
        score: Int,
        context: String = ""
    ) {
        self.id = id
        self.timestamp = timestamp
        self.type = type
        self.rawValue = rawValue
        self.score = score
        self.context = context
    }
}

struct RiskSnapshot {
    let timestamp: Date
    let totalScore: Int
    let level: RiskLevel
    let contributingEvents: [ThreatEvent]

    static var safe: RiskSnapshot {
        RiskSnapshot(timestamp: Date(), totalScore: 0, level: .safe, contributingEvents: [])
    }
}

extension Int {
    var toRiskLevel: RiskLevel {
        switch self {
        case 0...30:  return .safe
        case 31...65: return .medium
        case 66...85: return .high
        default:      return .critical
        }
    }
}

File: android/app/src/main/kotlin/com/shadowcell/scoring/ThreatEvent.kt
package com.shadowcell.scoring

import androidx.room.Entity
import androidx.room.PrimaryKey

enum class EventType {
    NETWORK_DOWNGRADE,      // 4G → 3G veya 3G → 2G
    NETWORK_UPGRADE_FAST,   // Çok hızlı upgrade (downgrade sonrası)
    SILENT_SMS_CANDIDATE,   // Delivery receipt ama inbox boş
    CELL_TOWER_ANOMALY,     // Bilinmeyen/zayıf tower handoff
    LOCATION_UPDATE_BURST,  // Çok sık location update
    SIGNAL_DROP_HANDOFF,    // Handoff sonrası sinyal düşüşü
    PAGING_ANOMALY,         // Beklenmedik device paging
}

enum class RiskLevel { SAFE, MEDIUM, HIGH, CRITICAL }

@Entity(tableName = "threat_events")
data class ThreatEvent(
    @PrimaryKey(autoGenerate = true) val id: Long = 0,
    val timestamp: Long = System.currentTimeMillis(),
    val type: EventType,
    val rawValue: String,       // Ham veri: "4G→3G", "CID:12345", vb.
    val score: Int,             // Bu event'in katkısı (0-100)
    val context: String = "",   // Ek bağlam: lokasyon, sinyal gücü, vb.
    val confirmed: Boolean = false,  // Kullanıcı "bu gerçekti" dedi mi
)

data class RiskSnapshot(
    val timestamp: Long = System.currentTimeMillis(),
    val totalScore: Int,
    val level: RiskLevel,
    val contributingEvents: List<ThreatEvent>,
    val windowSeconds: Int,
)

fun Int.toRiskLevel(): RiskLevel = when {
    this <= 30 -> RiskLevel.SAFE
    this <= 65 -> RiskLevel.MEDIUM
    this <= 85 -> RiskLevel.HIGH
    else -> RiskLevel.CRITICAL
}
